deepfence_agent/tools/apache/scope/probe/kubernetes/kubernetes_cluster.go

```diff
@@ -10,35 +11,75 @@ import (
 // KubernetesClusterResource represents a Kubernetes cluster
 type KubernetesClusterResource interface {
 	GetNode() report.TopologyNode
+	GetTopology() report.Topology
+	Stop()
 }
 
 type kubernetesCluster struct {
-	cloudProvider  string
-	cloudAccountID string
+	k8sTopology report.Topology
+	stopRefresh chan bool
+	sync.RWMutex
 }
 
 // NewKubernetesClusterResource creates a new Cluster node
 func NewKubernetesClusterResource() KubernetesClusterResource {
-	metadata := cloud_metadata.GetCloudMetadata()
-	return &kubernetesCluster{cloudProvider: metadata.CloudProvider, cloudAccountID: metadata.AccountID}
+	k8sCluster := kubernetesCluster{stopRefresh: make(chan bool)}
+	k8sCluster.cacheK8sTopology()
+	go k8sCluster.refresh()
+	return &k8sCluster
+}
+
+func (k *kubernetesCluster) Stop() {
+	k.stopRefresh <- true
+}
+
+func (k *kubernetesCluster) refresh() {
+	ticker := time.NewTicker(6 * time.Hour)
+	for {
+		select {
+		case <-ticker.C:
+			k.cacheK8sTopology()
+		case <-k.stopRefresh:
+			return
+		}
+	}
+}
+
+func (k *kubernetesCluster) cacheK8sTopology() {
+	k.Lock()
+	defer k.Unlock()
+
+	k.k8sTopology = report.MakeTopology()
+	node := k.GetNode()
+	//cloudProviderNodeId = node.Parents.CloudProvider
+	k.k8sTopology.AddNode(node)
+}
+
+func (k *kubernetesCluster) GetTopology() report.Topology {
+	k.RLock()
+	defer k.RUnlock()
+
+	return k.k8sTopology
 }
 
 func (k *kubernetesCluster) GetNode() report.TopologyNode {
+	cloudMetadata := cloud_metadata.GetCloudMetadata()
```

Comment:
We could avoid doing this if we know already we are not "private_cloud"

Reply:
Only after this we know whether we are in a known public cloud or not.